### PR TITLE
Classification Report (summary) #359

### LINKED ISSUE #359: Add support metric to ClassificationReport
Currently, the ClassificationReport omits support because it is difficult to put into the heatmap scale of (0.0, 1.0). We should still include it, however and _color_ it as the percent of the total number of records, while _labeling_ it with the actual support number.

### PR DESCRIPTION
**Pull request by ariley1472**
Fixes #359 

Initial Commit

Changes proposed in this pull request:

I want to make a pull request prior to making any significant changes. 
I'm trying to add another column to the Classification Report that shows the support. It looks like I just need to add the 'support' key value to the SCORE_KEYS and the functions within classification_report.py (including score, draw, and finalize). I still need to implement the actual drawing of the figure.

@DistrictDataLabs/team-oz-maintainers

### CODE REVIEW

**Comment by lwgray** [comment]
Thank you for joining us for the Yellowbrick Sprint at PyCon Cleveland. And also, thank you for submitting your first PullRequest - oh yeah! 

I really enjoyed working with you.  I hope to see you at the local python meetup.

We will review this asap. We have been quite busy with all the great contributions from sprinters like yourself.  In the meantime, feel free to work on another issue.  We hope you continue to contribute.

**Comment by lwgray** [comment]
@ariley1472 
We did a new yellowbrick release last night.. we are now at version 0.7  I performed an update on your branch thus you would need to pull from remote branch in order to bring your local repository in sync with remote (git pull origin develop)

**Comment by lwgray** [comment]
@ariley1472 I'm a bit disappointed that I have not received my invitation to Chocolate Club...

For this PR, there is only a minor change that needs to take place in order to merge the code .  If you like, I can commit changes to your Pull request.  I would just need your permission.  Below is a link to instructions on how to do this.

File: tests/test_classifier/test_classification_report.py [hunk 1]
```diff
@@ -141,6 +146,9 @@ def test_pandas_integration(self):
             }, 'f1': {
                 'unoccupied': 0.9800031994880819,
                 'occupied': 0.9366447034972124
+            }, 'support': {
+                'occupied': 0.22519455252918288,
+                'unoccupied': 0.77480544747081714
             }}
 
     @pytest.mark.skip(reason="requires random state in quick method")
```
Comment (bbengfort) [inline tests/test_classifier/test_classification_report.py]: @lwgray would it be possible for you to add the random state in the quick method to repair this test?

File: yellowbrick/classifier/classification_report.py [hunk 2]
```diff
@@ -101,26 +102,36 @@ def score(self, X, y=None, **kwargs):
...
+
+        # Calculate the percentage for the support metric
+        self.support_score = scores[-1]
+        support_percent = self.support_score / (sum(self.support_score))
+
+        scores = map(lambda s: dict(zip(self.classes_, s)), scores)
         self.scores_ = dict(zip(SCORES_KEYS, scores))
 
+        # Change the support score from the actual support value to the percent
+        # value to be used in the Classification Report.
```
Comment (bbengfort) [inline yellowbrick/classifier/classification_report.py]: Thank you for the excellent code documentation!

**Comment by ariley1472** [comment]
Hi @lwgray. My apologies for being MIA. I've been really busy with work. I'll grant you permission to commit changes. Thanks for your patience!

File: yellowbrick/classifier/classification_report.py [hunk 3]
```diff
@@ -134,6 +145,11 @@ def draw(self):
                 value = cr_display[x,y]
                 svalue = "{:0.3f}".format(value)
 
+                # change the svalue for support (when y == 3) because we want
+                # to label it as the actual support value, not the percentage
```
Comment (bbengfort) [inline yellowbrick/classifier/classification_report.py]: After seeing some of the example images, I'd actually recommend that we make this an option for the user in `__init__` - I'm thinking the user should specify:

1. Show support (or don't show support for backward compatibility) 
2. Show support as percent 
3. Show support as real number 

We can do this with a single argument:

`support="count"` 

Where support can be None, False, True, or one of "percent", "count". If None or False, do not display support; if "count" displays the actual count, otherwise display percent. 

If not in this PR, we could open a high priority issue for this.

File: docs/api/classifier/classification_report.rst [hunk 4]
```diff
@@ -56,8 +56,9 @@ The metrics are defined in terms of true and false positives, and true and false
 **f1 score**
     The F\ :sub:`1` score is a weighted harmonic mean of precision and recall such that the best score is 1.0 and the worst is 0.0. Generally speaking, F\ :sub:`1` scores are lower than accuracy measures as they embed precision and recall into their computation. As a rule of thumb, the weighted average of F\ :sub:`1` should be used to compare classifier models, not global accuracy.
 
-.. caution::
-    Support is omitted from the classification metrics because it is difficult to scale this on a heatmap. In a future release we will add a support feature that describes the support as the percentage of the total number of instances.
+
+**support**
```
Comment (bbengfort) [inline docs/api/classifier/classification_report.rst]: Recommend a bit more detail:

> Support is the number of actual occurrences of the class in the specified dataset. Imbalanced support in the training data may indicate structural weaknesses in the reported scores of the classifier and could indicate the need for stratified sampling or rebalancing. Support doesn't change between models but instead diagnoses the evaluation process.

**Comment by bbengfort** [comment]
@ariley1472 great to hear from you! No worries - we completely understand you're being busy - @lwgray can wrap up the small stuff; thank you so much for your contribution!

**Comment by lwgray** [comment]
I've Updated this PR by
1. Adding random state parameter to classification report quick_method
2. Increasing tolerance for 'assert_similar_test' in classification_report test
3. Expanding explanation of Support in docs
4. Adding import statement 'from __future__ import division'

If applied the above changes will
1. Allow for classification report's quickmethod test to pass
2. Improve new classification report 'support' documentation
3. Will correct Travis python 2.7 tests failure that was due to how python 2
   performs division


image

File: tests/test_classifier/test_classification_report.py [hunk 5]
```diff
@@ -128,9 +133,9 @@ def test_pandas_integration(self):
         viz.fit(X_train, y_train)
         viz.score(X_test, y_test)
 
-        self.assert_images_similar(viz, tol=0.1)
+        self.assert_images_similar(viz, tol=43.0)
```
Comment (bbengfort) [inline tests/test_classifier/test_classification_report.py]: This is a fairly big jump in tolerance; could we also create an issue to review why these high tolerances are required? Maybe just note the classification report in #379 has pretty high tolerances and they should be reviewed.

File: yellowbrick/classifier/classification_report.py [hunk 6]
```diff
@@ -4,6 +4,7 @@
 # Author:   Rebecca Bilbro <rbilbro@districtdatalabs.com>
 # Author:   Benjamin Bengfort <bbengfort@districtdatalabs.com>
 # Author:   Neal Humphrey
+# Author:   Allyssa Riley
```
Comment (bbengfort) [inline yellowbrick/classifier/classification_report.py]: @lwgray you could add your name here as well!

File: tests/test_classifier/test_classification_report.py [hunk 7]
```diff
@@ -141,9 +146,11 @@ def test_pandas_integration(self):
             }, 'f1': {
                 'unoccupied': 0.9800031994880819,
                 'occupied': 0.9366447034972124
+            }, 'support': {
+                'occupied': 0.22519455252918288,
+                'unoccupied': 0.77480544747081714
             }}
 
-    @pytest.mark.skip(reason="requires random state in quick method")
```
Comment (bbengfort) [inline tests/test_classifier/test_classification_report.py]: Thank you for removing this skip! All the tests pass, possibly because of the increased tolerance; but should we also add the random state here as well?

File: yellowbrick/classifier/classification_report.py [hunk 8]
```diff
@@ -14,11 +15,12 @@
 """
 Visual classification report for classifier scoring.
 """
-
+ 
 ##########################################################################
 ## Imports
 ##########################################################################
 
+from __future__ import division
```
Comment (bbengfort) [inline yellowbrick/classifier/classification_report.py]: 👍

File: yellowbrick/classifier/classification_report.py [hunk 9]
```diff
@@ -215,7 +232,8 @@ def classification_report(model, X, y=None, ax=None, classes=None, **kwargs):
     visualizer = ClassificationReport(model, ax, classes, **kwargs)
 
     # Create the train and test splits
-    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2,
+                                                        random_state=42)
```
Comment (bbengfort) [inline yellowbrick/classifier/classification_report.py]: Ah, I see what happened - I meant add random state as an argument to this function (Default=None), then pass it to the `train_test_split`function, that way the user can supply the random state if required (e.g. for testing).

File: tests/test_classifier/test_classification_report.py (hunk 5 above)
Comment (lwgray) [reply tests/test_classifier/test_classification_report.py]: @ndanielsen 
These tolerances were set at sprints and I didn't closely monitor them thus they might be artificially inflated.

File: yellowbrick/classifier/classification_report.py (hunk 6 above)
Comment (lwgray) [reply yellowbrick/classifier/classification_report.py]: sure!

File: tests/test_classifier/test_classification_report.py (hunk 5 above)
Comment (bbengfort) [reply tests/test_classifier/test_classification_report.py]: Awesome, the issue you created is perfect.

File: tests/test_classifier/test_classification_report.py [hunk 10]
```diff
@@ -154,9 +161,10 @@ def test_quick_method(self):
         )
 
         _, ax = plt.subplots()
-        classification_report(DecisionTreeClassifier(), X, y, ax=ax)
+        classification_report(DecisionTreeClassifier(), X, y,
+                              ax=ax, random_state=42)
```
Comment (bbengfort) [inline tests/test_classifier/test_classification_report.py]: 👍

**Comment by ariley1472** [comment]
Thank you guys for working on this! Hopefully when work slows down, I'll have more time to work on other projects for YellowBrick! :D

**Comment by lwgray** [comment]
@ariley1472 No Worries... We will make sure to add your name as a contributor.  We also have a host of novice/intermediate issues you can work on.  😄 
https://github.com/DistrictDataLabs/yellowbrick/issues?q=is%3Aopen+is%3Aissue+label%3A%22level%3A+novice%22
